repos/backend/endpoints/files.js

```diff
@@ -43,7 +45,98 @@ const deleteFile = (app, config) => async (req, res) => {
   }
 }
 
+/**
+ * Recursively checks to find the parent node for a given item
+ * if a parent exists, it will add it as a child
+ * @param {Array<Object>} nodes 
+ * @param {string} parentPath 
+ * @param {Object} newItem
+ * 
+ * @returns {Boolean} - whether a parent node exists and push was successful
+ */
+const parentNodeExists = (nodes, parentPath, newItem) => {
+  const found = nodes.find((node) => {
+    return node.fullPath === parentPath
+      ? Boolean(node.children.push(newItem))
+      : node.children.length && parentNodeExists(node.children, parentPath, newItem)
+  })
+
+  return Boolean(found)
+}
+
+/**
+ * Gets the metadata of a path from the local filesystem
+ * @param {string} path - full path to the folder or file i.e '/keg/tap/tests/bdd/features'
+ * 
+ * @returns {Object} - Meta data containing {name, parent, type ( folder || file )} properties
+ */
+const getPathMeta = path => {
+  const pathSplit = path.split('/')
+  return {
+    name: pathSplit.pop(),
+    parent: pathSplit.join('/'),
+    type: fs.lstatSync(path).isDirectory() ? 'folder' : 'file',
+  }
+}
+
+/**
+ * Transforms the paths string to a specific data object
+ * @param {Array<string>} paths - full paths to the folder or file i.e '/keg/tap/tests/bdd/features'
+ * 
+ * @returns {Array<Object>} - each object has the form: 
+ *                            {id, fullPath, children: [], isModified}
+ */
+const generateTree = (paths) => {
+   /**
+   * 1. create new object for each 'path' item
+   * 2. if the parent path of current 'path' item exists, add it as the child
+   */
+  return paths.reduce((nodes, path) => {
+    // Get the meta data for this path
+    const { parent, ...pathMeta } = getPathMeta(path)
+    
+    // Ignore hidden files that start with a .
+    if (pathMeta.type === 'file' && pathMeta.name.startsWith('.')) return nodes
+  
+    const node = {
+      id: path,
+      children: [],
+      fullPath: path,
+      isModified: false,
+      ...pathMeta,
+    }
+
+    // either push the node or add it to an existing node.children
+    ;(!nodes.length || !parentNodeExists(nodes, parent, node)) &&
+      nodes.push(node)
+
+    return nodes
+  }, [])
+
+}
+
+/**
+ * iterates through the mounted test root folder and returns a tree like structure of all the folders/files
+ * @param {Object} app 
+ * @param {object} config - shared config
+ */
+const getTree = (app, config) => async (req, res) => {
+  try {
+    const { testsRoot } = config.paths
+    const meta = await getFolderContentSync(testsRoot, {
```

Comment:
I thought I mentioned this, but maybe I didn't.
* When using `Node.js` as a backend for API's we almost never want to do anything **synchronous**
* This causes the API to be locked up, or unresponsive while it's executing the `synchronous` process.
  * You can also google it for more infor, but here's a couple links to get you started
    * https://rclayton.silvrback.com/avoid-synchronous-functions-in-node-js
    * https://www.oreilly.com/library/view/isomorphic-javascript-web/9781785889769/12a16110-8751-422a-bf82-63a9f9a4f9e3.xhtml
    * https://www.google.com/search?q=node.js+don%27t+use+synchronous+functions&oq=node.js+don%27t+use+synchronous+functions&aqs=chrome..69i57j0i333.5912j0j7&sourceid=chrome&ie=UTF-8

* Please update any of these calls that are  `synchronous`, to be `asynchronous`, a.k.a `Non-Blocking`

Reply:
ok thanks. I updated the sync `fs` methods to be the async version limboified